src/progress.rs

```diff
@@ -105,41 +105,37 @@ impl ProgressSet {
     }
 
     /// Returns an iterator across all the nodes and their progress.
-    pub fn iter(&self) -> Chain<Iter<u64, Progress>, Iter<u64, Progress>> {
+    pub fn iter(&self) -> impl Iterator<Item = (&u64, &Progress)> {
         self.voters.iter().chain(&self.learners)
     }
 
     /// Returns a mutable iterator across all the nodes and their progress.
-    pub fn iter_mut(&mut self) -> Chain<IterMut<u64, Progress>, IterMut<u64, Progress>> {
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = (&u64, &mut Progress)> {
         self.voters.iter_mut().chain(&mut self.learners)
     }
 
     /// Adds a voter node
-    ///
-    /// # Panics
-    ///
-    /// Panics if the node already has been added.
-    pub fn insert_voter(&mut self, id: u64, pr: Progress) {
-        if self.learners.contains_key(&id) {
-            panic!("insert voter {} but already in learners", id);
+    pub fn insert_voter(&mut self, id: u64, pr: Progress) -> Result<(), Error> {
+        if self.voters.contains_key(&id) {
+            Err(Error::Exists(id, "voters"))?
         }
-        if self.voters.insert(id, pr).is_some() {
-            panic!("insert voter {} twice", id);
+        if self.learners.contains_key(&id) {
+            Err(Error::Exists(id, "learners"))?;
```

Comment:
any different to panic outside?

Reply:
The error message will be different now. Otherwise this is just moving the error up a level so these functions can be called without worrying about panic.